lib/nrf_modem_lib/trace_backends/flash/flash.c

```diff
@@ -0,0 +1,392 @@
+
+	memcpy(&stream_buf[stream_buf_written], data, append_len);
+
+	stream_buf_written += append_len;
+	trace_bytes_written += append_len;
+
+	return append_len;
+}
+
+static int fcb_walk_callback(struct fcb_entry_ctx *loc_ctx, void *arg)
+{
+	if ((loc_ctx->loc.fe_sector == sector) && (loc_ctx->loc.fe_elem_off < loc.fe_elem_off)) {
+		return 0;
+	}
+
+	trace_bytes_written -= loc_ctx->loc.fe_data_len;
+	return 0;
+}
+
+static int buffer_flush(void)
+{
+	int err;
+	struct fcb_entry loc_flush;
+
+	if (!is_initialized) {
+		return -EPERM;
+	}
+
+	if (!stream_buf_written) {
+		return -ENODATA;
+	}
+
+	err = fcb_append(&trace_fcb, stream_buf_written, &loc_flush);
+	if (err) {
+		if (IS_ENABLED(CONFIG_NRF_MODEM_TRACE_FLASH_NOSPACE_ERASE_OLDEST)) {
+			/* Find the number of trace bytes in oldest sector (that is not read). */
+
+			/* Get first sector in FCB */
+			loc_flush.fe_sector = 0;
+			loc_flush.fe_elem_off = 0;
+			err = fcb_getnext(&trace_fcb, &loc_flush);
+
+			/* Walk sector to remove unread trace data from count. */
+			err = fcb_walk(&trace_fcb, loc_flush.fe_sector, fcb_walk_callback, NULL);
+			if (err) {
+				LOG_ERR("fcb_walk failed, err %d", err);
+				return err;
+			}
+
+			/* Erase the oldest sector and append again. */
+			err = fcb_rotate(&trace_fcb);
+			if (err) {
+				LOG_ERR("fcb_rotate failed, err %d", err);
+				return err;
+			}
+			err = fcb_append(&trace_fcb, stream_buf_written, &loc_flush);
+		}
+
+		if (err) {
+			LOG_ERR("fcb_append failed, err %d", err);
+			return -ENOSPC;
+		}
+	}
+
+	err = flash_area_write(
+		trace_fcb.fap, FCB_ENTRY_FA_DATA_OFF(loc_flush), stream_buf, stream_buf_written);
+	if (err) {
+		LOG_ERR("flash_area_write failed, err %d", err);
+		return err;
+	}
+
+	err = fcb_append_finish(&trace_fcb, &loc_flush);
+	if (err) {
+		LOG_ERR("fcb_append_finish failed, err %d", err);
+		return err;
+	}
+
+	stream_buf_written = 0;
+
+	return 0;
+}
+
+static int trace_flash_erase(void)
+{
+	int err;
+
+	LOG_INF("Erasing external flash");
+
+	err = flash_area_erase(modem_trace_area, 0, TRACE_SIZE);
+	if (err) {
+		LOG_ERR("flash_erase error: %d", err);
+		return err;
+	}
+
+	return err;
+}
+
+static const uint32_t magic_val = 0x152ac523;
+
+int trace_backend_init(trace_backend_processed_cb trace_processed_cb)
+{
+	int err;
+	const struct flash_parameters *fparam;
+
+	if (trace_processed_cb == NULL) {
+		return -EFAULT;
+	}
+
+	trace_processed_callback = trace_processed_cb;
+
+	err = flash_area_open(FIXED_PARTITION_ID(MODEM_TRACE), &modem_trace_area);
+	if (err) {
+		LOG_ERR("flash_area_open error:  %d", err);
+		return -ENODEV;
+	}
+
+	err = flash_area_has_driver(modem_trace_area);
+	if (err == -ENODEV) {
+		LOG_ERR("flash_area_has_driver error: %d\n", err);
+	}
+
+	flash_dev = flash_area_get_device(modem_trace_area);
+	if (flash_dev == NULL) {
+		LOG_ERR("Failed to get flash device\n");
+		return -ENODEV;
+	}
+
+	/* After a cold boot the magic will contain random values. */
+	if (magic != magic_val) {
```

Comment:
Not sure whether we need this.
There are three cases,
- nothing is in flash
- flash contains fcb data (modem traces)
- flash contains something other than traces (fcb not installed, pre-existing data?)

In the first two cases, we don't need to erase anything.
In the last case, what will happen during `fcb_init` ?
If that's going to return an error, can we simply erase then and retry?

Reply:
It looks like the `stream_buf` is preserved on soft-boot, but shouldn't `stream_buf` be written to flash before rebooting? Do we need to preserve `stream_buf`?

Reply:
I've rewritten this today so that the buffer is flushed when the trace backend is deinitialized.

Reply:
However, we still need to keep track of how much we have read (which I added to `__noinit` today).